### ISSUE TITLE
Implement Camera Handler

### ISSUE THREAD

**Issue by ClayJay3** [MissouriMRDT/Autonomy_Software#110]
- [x] Create base camera class.
- [x] Implement BasicCam
- [x] Implement ZedCam (*almost there*)
- [x] #124
- [x] #123
- [x] Create CameraHandler class with multithreading abilities (*partially written*)

**Comment by ClayJay3** [comment]
shooo buddy, multithreading. After writing 1000 lines of code, it turns out to be kinda trash. I suppose some things really can't be done perfectly the first time. Going to rewrite some of the portions that deal with threaded frame grabbing and retrieval so that they use double buffers with a condition variable. This should fix the deadlocking issues. Proper mutexes, unique, and shared locks will still be used to ensure that multiple threads can access the same copy of an image until a new one is grabbed.

The python codebase implementation is definitely suboptimal.

**Comment by ClayJay3** [comment]
got something to work, it's knarly. I was able to get two copies of the normal frame without any noticeable drops in FPS for the zed grabber thread. Used a threadpool to 'blast' copy queued frames all at once.

Each call to GrabFrame() creates a struct var that stores a condition variable, mutex, and cv:Mat&. The struct is initialized and a reference to it is put into a queue. Then the GrabFrame() method does a wait() on the condition variable. This can be called as many times as needed and each call will add a struct storing a cv::Mat& to the queue and then block. 

In the threaded zed grabber method, the thread waits to acquire a lock from the queue. The zed grabber thread is given locking priority due to the mutexes and locks I used. When the lock is acquired, it sends copies of a method as tasks to the pool. Each thread in the threadpool does a quick nested lock and pop out of the queue to get the struct storing the cv::Mat&. Then each thread in the pool copies the data from the m_slFrame member variable and signals the condition variable to notify any other threads in wait() to stop waiting. At this point it doesn't matter how long it takes for all the frames to copy. As soon as the thread in the threadpool signals, the GrabFrame() function in the other thread immediately returns the copied frame.

Then once all the threads in the pool are done, the zed grabber thread joins the threadpool and does the loop over again.
